### ISSUE TITLE
Use the hash of the article title to provide the article url

### ISSUE THREAD

**Issue by you-hengh** [vuejs/vitepress#3326]
### Is your feature request related to a problem? Please describe.

In the process of daily use, non-native English speakers will use their native language to write article titles when writing articles.

However, the url obtained by such an article after publication will be encoded, resulting in the blog's post address becoming very long

### Describe the solution you'd like

Here's an example.

<img width="772" alt="image" src="https://github.com/vuejs/vitepress/assets/34816426/e5a30849-286c-472c-9c85-aa5c5a873113">

After copying the url, the URL becomes like this:

> https://blog.jimmyxuexue.top/front-end/%E6%8A%98%E8%85%BE/%E5%AE%9E%E7%8E%B0%E9%82%AE%E7%AE%B1%E8%BE%85%E5%8A%A9%E7%99%BB%E5%BD%95%E5%8A%9F%E8%83%BD.html

This is very inelegant in the process of sharing articles, the address is too messy, and some editing boxes will limit the number of characters.

### Describe alternatives you've considered

I hope vitepress can provide a configuration item that automatically turns the article address into a hash value of the article title (preferably to control the length, such as 6 bits, 10 bits, etc.)

So even if I change the content of the article, the title of the article will not change, and the address of the article will not change

### Additional context

Original url:

> https://blog.jimmyxuexue.top/front-end/%E6%8A%98%E8%85%BE/%E5%AE%9E%E7%8E%B0%E9%82%AE%E7%AE%B1%E8%BE%85%E5%8A%A9%E7%99%BB%E5%BD%95%E5%8A%9F%E8%83%BD.html

Changed url:

> https://blog.jimmyxuexue.top/front-end/jl34n23rnl.html

Change only the article title, keep the folder path

### Validations

- [X] Follow our [Code of Conduct](https://vuejs.org/about/coc.html)
- [X] Read the [docs](https://vitepress.dev).
- [X] Read the [Contributing Guidelines](https://github.com/vuejs/vitepress/blob/main/.github/contributing.md).
- [X] Check that there isn't already an issue that asks for the same feature to avoid creating a duplicate.

**Comment by Justin3go** [comment]
我也有这个问题，不过简单地使用内存Map实现了一个静态短链接服务如，[justin3go.com/s?u=590f2bc50aa](justin3go.com/s?u=590f2bc50aa)

也许还可以尝试一下市面上的一些短链接服务

**Comment by you-hengh** [comment]
> 我也有这个问题，不过简单地使用内存Map实现了一个静态短链接服务如，[justin3go.com/s?u=590f2bc50aa](justin3go.com/s?u=590f2bc50aa)
> 
> 也许还可以尝试一下市面上的一些短链接服务

并不方便,而且那些公用短链服务可能被腾讯标红或者拦截
Not convenient, and those public short URL services may be red or blocked by Tencent

**Comment by brc-dd** [comment]
Ah, so it can't be an actual hash (since you want it to remain constant across edits), it should be a random string which is stored in some file(s) (tracked by your vcs) so that it doesn't change across builds.

The most feasible way I see is to just store this in frontmatter of each post:

```md
---
slug: V1StGXR8
---
```

And in `config.ts` generate [rewrites](https://vitepress.dev/guide/routing#route-rewrites) using fast-glob + gray-matter

**Comment by you-hengh** [comment]
> Ah, so it can't be an actual hash (since you want it to remain constant across edits), it should be a random string which is stored in some file(s) (tracked by your vcs) so that it doesn't change across builds.啊，所以它不能是一个实际的散列（因为你希望它在编辑过程中保持不变），它应该是一个随机字符串，存储在某些文件中（由你的 vcs 跟踪），这样它就不会跨构建的变化。
> 
> The most feasible way I see is to just store this in frontmatter of each post:我认为最可行的方法是将其存储在每个帖子的 frontmatter 中：
> 
> ```
> ---
> slug: V1StGXR8
> ---
> ```
> 
> And in `config.ts` generate [rewrites](https://vitepress.dev/guide/routing#route-rewrites) using fast-glob + gray-matter并在 `config.ts` 中使用 fast-glob + grey-matter 生成重写

You may not understand what I mean, I mean hashing the title of the article, if I have an article that has been published, I find something wrong with the content and modify it, but I will not change the title of the article (rarely will I change the title),

To clarify again, I just said to hash the title of the article, not the entire content of the article. My expression has gone through the translator, and I don't know if my description is accurate.

**Comment by brc-dd** [comment]
Ah, okay. Now I get it. Earlier I thought you wanted hash of whole document.

**Comment by wstee** [comment]
我想用frontmatter中的属性来生成路由，能实现吗？

**Comment by you-hengh** [comment]
> Ah, okay. Now I get it. Earlier I thought you wanted hash of whole document.

hi, I want to ask, is there any progress on this issue now? I try to provide my blog address to others, but obviously the url address is very long, which is too ugly

https://github.com/vuejs/vitepress/discussions/1646#discussioncomment-9841679

**Comment by brc-dd** [comment]
I have one question, even if we do rewrites on pages, how would the links to headings look? They will too get url encoded 👀

**Comment by you-hengh** [comment]
> I have one question, even if we do rewrites on pages, how would the links to headings look? They will also get url encoded 👀

Indeed, this is a topic worth discussing, because in addition to the article title, the URL may also include Chinese-named categories, as shown in the figure:

![2024-06-23_14-24-23](https://github.com/vuejs/vitepress/assets/34816426/9f4774df-eead-4171-8973-ef190de56573)

Not only that, but also consider <a>: The Anchor element:
![2024-06-23_14-39-30](https://github.com/vuejs/vitepress/assets/34816426/84991025-fa03-4247-8424-342911ea13f8)

This may increase the difficulty and require enough patience to solve the problem

**Comment by you-hengh** [comment]
I suddenly had a fantastic idea, maybe we don't have to think about how to shorten the url, we just need to add a button or a element to the article page, display the copy article address, and write the url with Chinese to the user's clipboard. 

> https://bddxg.top/article/note/工程化/脚手架/Vite笔记.html

@brc-dd

**Comment by brc-dd** [comment]
Any good UI idea? 👀 I think below outline might be a good place to keep it, but not sure how it would look on smaller viewports.

**Comment by you-hengh** [comment]
> Any good UI idea? 👀 I think below outline might be a good place to keep it, but not sure how it would look on smaller viewports. Also, it won't look nice if outline is long. Some other place maybe.

Unfortunately, I don't have much experience with UI. Perhaps a codepen search would be a good idea, for example:

https://codepen.io/search/pens?q=share+button

If it's a button layout, I thought it would be a good place to be on the right side of the title (probably dealing with super long title wrapping issues)

![2024-06-23_16-18-17](https://github.com/vuejs/vitepress/assets/34816426/e6073e79-6730-4b45-93a6-71db520f9edd)

**Comment by clicelee** [comment]
Hi, I've been following this discussion with great interest. The idea of a "decoded URL copy button" to address the long URL issue is a good approach for improving user experience, and I support it.

In addition to this, I'd like to offer a suggestion regarding another important challenge that can arise, particularly when managing multilingual documentation: ensuring consistency of internal anchor links (#fragments) across different language versions of the same document.

VitePress already provides an excellent feature for specifying custom anchors for Markdown headers using the {#custom-anchor} syntax.
[VitePress Docs: Header Anchors - Custom anchors](https://vitepress.dev/guide/markdown#header-anchors)

Proposed Idea:

I propose introducing a system that enforces matching custom anchors for headers within documents that share the same filename across different locales (e.g., en/guide.md and ko/guide.md).

For instance, if en/guide.md has the following headers and custom anchors:
```
## Installation {#installation}
## Configuration {#configuration}
```

Then the corresponding ko/guide.md (or guide.md in any other language) would be required to use the same custom anchors in the same order and with the same names:
```
## 설치 {#installation}       (Korean for "Installation")
## 설정 {#configuration}       (Korean for "Configuration")
```

Expected Benefits:

Accurate Internal Linking in Multilingual Contexts: This would ensure that anchor links point to the correct section even when the user switches languages (e.g., /en/guide#installation correctly links to the same conceptual section as /ko/guide#installation).

Improved Maintainability: It would help prevent errors caused by missing or mismatched anchors during translation and content updates.

Potential SEO Improvement: A consistent URL structure for anchored content could also positively impact SEO.

Possible Implementation Approaches:

This could be implemented as a check during the build process, or via Git pre-push/pre-commit hooks, which would compare custom anchors across multilingual files and raise warnings or errors if inconsistencies are found, prompting correction.

A VitePress plugin or a separate validation tool could also be developed for this purpose.

While this idea is distinct from the "URL copy button" currently under discussion, I believe it could significantly enhance the structural integrity, overall usability, and maintainability of multilingual documentation. I think this aspect is also worth discussing.

I'd appreciate any thoughts on this!

**Comment by clicelee** [comment]
And just to add, if this specific discussion about enforcing custom anchor consistency across multilingual documents seems to warrant its own thread, I'd be happy to create a new issue dedicated to it.

I'm passionate about solving this problem. In fact, in a project where [toss](https://github.com/toss) uses VitePress [toss/frontend-fundamentals#160](https://github.com/toss/frontend-fundamentals/pull/160) - link to your PR for context, if applicable), I initially created a `custom slugify function` to manage similar needs around generating consistent and predictable anchor links for Korean content. However, I believe addressing this challenge within VitePress itself, perhaps as an optional feature or a plugin, would be a more robust and beneficial solution for the wider community.

I'm keen to contribute to finding a good solution here if there's interest.
